### ISSUE TITLE
Debug runtime assertion failure on ARM32

### ISSUE THREAD

**Issue by jmid** [ocaml/ocaml#12267]
While hacking CI tests for domainslib [I discovered a test that would trigger an assertion failure under the debug runtime on ARM32 only](https://github.com/ocaml-multicore/domainslib/pull/112#issuecomment-1562716601). I've confirmed locally on a Raspberry Pi 2 with both `5.0.0` and `5.1.0~alpha1`.

To recreate:

- fork https://github.com/jmid/domainslib/tree/lu-decomposition-abort
- `opam install domain-local-await lockfree`
- `dune exec test/LU_decomposition_multicore.exe`

```
Done: 38% (17/44, 27 left) (jobs: 0)### OCaml runtime: debug mode ###
[00] file runtime/fiber.c; line 182 ### Assertion failed: Stack_high(stack) - Stack_base(stack) == wosize || Stack_high(stack) - Stack_base(stack) == wosize + 1
Aborted
```

I am now building an opam switch for `5.1.0+trunk` (which follows the ocaml/ocaml `5.1` branch) after yesterday's many cherry picks, to see whether any of them should fix it. I'll share the results here once it is ready (it is slow... :sweat_smile:)

**Comment by jmid** [comment]
I can confirm the assertion failure with the `5.1.0+trunk` switch too.
The version installed that way identifies as `OCaml version 5.1.0+dev5-2023-04-14` though...
I may therefore try a local opam switch with the `5.1` branch too.

**Comment by Octachron** [comment]
`5.1.0+dev5-2023-04-14` is the current dev version for the branch 5.1 .

**Comment by xavierleroy** [comment]
I didn't write this code, but this assertion smells bad.  A few lines before it says "Ensure 16-byte alignment because some architectures require it", and then the assertion checks that something is 0 to 1 machine word away from something else.  On a 32-bit machine like ARM, 16-byte alignment means 0 to 3 machine words, not 0 to 1.

@stedolan : you wrote this assertion; what do you think?

**Comment by ghost** [comment]
There are two problems here: the assertion is too strict since alignment may cause more than one extra word to be added, and the alignment computation was wrong and could actually cause the stack to end up smaller than requested.

A PR attempting to fix these has been submitted [here](https://github.com/ocaml/ocaml/pull/12434).

**Comment by gasche** [comment]
This should have been fixed by #12434 from @dustanddreams.
